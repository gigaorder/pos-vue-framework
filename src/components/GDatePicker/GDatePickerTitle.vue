<script>
  import PickerButton from '../../mixins/picker-button';
  import { computed, createElement as h, reactive, watch } from '@vue/composition-api';

  export default {
    name: 'GDatePickerTitle',
    props: {
      date: {
        type: String,
        default: '',
      },
      disabled: Boolean,
      readonly: Boolean,
      selectingYear: Boolean,
      value: {
        type: String
      },
      year: {
        type: [Number, String],
        default: '',
      },
      yearIcon: {
        type: String,
      },
    },
    setup(props, context) {
      // mixins
      const { genPickerButton } = PickerButton(props, context)

      // data
      const state = reactive({isReversing: false})
      // computed
      let computedTransition = computed(() => {
        return state.isReversing ? 'picker-reverse-transition' : 'picker-transition'
      })

      // watch
      watch(() => props.value, (val, prev) => {
        state.isReversing = val < prev
      }, { lazy: true })

      // methods
      function genYearIcon ()/*: VNode */{
        return h('div' /*TODO: VIcon*/, {
          props: {
            // dark: true,
          },
        }, props.yearIcon)
      }
      function getYearBtn ()/*: VNode*/ {
        return genPickerButton('selectingYear', true, [
          String(props.year),
          props.yearIcon ? genYearIcon() : null,
        ], false, 'g-date-picker-title__year')
      }
      function genTitleText ()/*: VNode */{
        return h('transition', {
          props: {
            name: computedTransition.value,
          },
        }, [
          h('div', {
            domProps: { innerHTML: props.date || '&nbsp;' },
            // key: props.value,
          }),
        ])
      }
      function genTitleDate ()/*: VNode*/ {
        return genPickerButton('selectingYear', false, [genTitleText()], false, 'g-date-picker-title__date')
      }

      return () => {
        return h('div', {
          staticClass: 'g-date-picker-title',
          class: {
            'g-date-picker-title--disabled': props.disabled,
          },
        }, [
          getYearBtn(),
          genTitleDate(),
        ])
      }
    }
  }
</script>
<style scoped>
</style>
