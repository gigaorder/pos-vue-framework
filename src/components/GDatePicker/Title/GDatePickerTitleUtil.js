export const EVENT_NAMES = {
  UPDATE_SELECTING_YEAR: 'update:selecting-year'
}

export function getYearButtonClickHandler(props, context) {
  return (e) => {
    e.stopPropagation()
    if (props.selectingYear || props.readonly)
      return
    context.emit(EVENT_NAMES.UPDATE_SELECTING_YEAR, true)
  }
}
