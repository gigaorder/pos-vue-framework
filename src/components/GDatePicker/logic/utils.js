/**
 * Filled '0' on the left of the input if input length is less than target length
 *
 * @param value { number | string } input value
 * @param targetLength { number } the length of output string
 * @returns {string} Filled string
 */
export const pad = (value, targetLength = 2) => value.toString().padStart(targetLength, '0')

// Adds leading zero to month/day if necessary, returns 'YYYY' if type = 'year',
// 'YYYY-MM' if 'month' and 'YYYY-MM-DD' if 'date'
export function sanitizeDateString(dateString/*: string*/, type/*: DATE_PICKER_TYPE*/)/*: string*/ {
  const [year, month = 1, date = 1] = dateString.split('-')
  return `${year}-${pad(month)}-${pad(date)}`.substr(0, { date: 10, month: 7, year: 4 }[type])
}
