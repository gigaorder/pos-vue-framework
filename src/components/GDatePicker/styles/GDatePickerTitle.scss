.g-date-picker-title {
  display: flex;
  flex-direction: column;
  line-height: 1;

  &__year {
    display: inline-flex;
    font-size: 14px;
    font-weight: 500;
    margin-bottom: 8px;
  }

  &__date {
    font-size: 34px;
    text-align: left;
    font-weight: 500;
    position: relative;
    overflow: hidden;
    padding-bottom: 8px;
    margin-bottom: -8px;
  }

  &--disabled {
    pointer-events: none;
  }
}
