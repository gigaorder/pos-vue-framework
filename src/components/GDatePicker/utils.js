/**
 * Make ISO string
 * @param dateString
 * @returns {string}
 */
const makeIsoString = (dateString) => {
  const [year, month, date] = dateString.trim().split(' ')[0].split('-')
  return [pad(year, 4), pad(month || 1), pad(date || 1)].join('-')
}

/**
 *
 * @param locale {string | undefined}
 * @param options {Intl.DateTimeFormatOptions}
 * @param substrOptions {SubstrOptions = { start: 0, length: 0 }}
 * @returns {function(string): string | undefined}
 */
export function createNativeLocaleFormatter (locale, options, substrOptions = { start: 0, length: 0 }) {
  try {
    const intlFormatter = new Intl.DateTimeFormat(locale || undefined, options)
    return (dateString) => intlFormatter.format(new Date(`${makeIsoString(dateString)}T00:00:00+00:00`))
  } catch (e) {
    return (substrOptions.start || substrOptions.length)
        ? (dateString) => makeIsoString(dateString).substr(substrOptions.start || 0, substrOptions.length)
        : undefined
  }
}



/**
 * @param {String} value YYYY-MM format
 * @param {Number} sign -1 or +1
 */
export const monthChange = (value, sign) => {
  const [year, month] = value.split('-').map(Number)

  if (month + sign === 0) {
    return `${year - 1}-12`
  } else if (month + sign === 13) {
    return `${year + 1}-01`
  } else {
    return `${year}-${pad(month + sign)}`
  }
}

/**
 * Filled '0' on the left of the input if input length is less than target length
 *
 * @param value { number | string } input value
 * @param targetLength { number } the length of output string
 * @returns {string} Filled string
 */
export const pad = (value, targetLength = 2) => value.toString().padStart(targetLength, '0')


// --
export const DAYS_IN_MONTH = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
export const DAYS_IN_MONTH_LEAP = [0, 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

function isLeapYear (year) {
  return ((year % 4 === 0) && (year % 100 !== 0)) || (year % 400 === 0)
}

export function daysInMonth (year, month) {
  return isLeapYear(year) ? DAYS_IN_MONTH_LEAP[month] : DAYS_IN_MONTH[month]
}

export const TRANSITION_NAMES = {
  TAB: 'tab-transition',
  REVERSE_TAB: 'tab-reverse-transition',
  PICKER: 'picker-transition'
}

export const SANITY_TYPE = {
  DATE: 'date',
  MONTH: 'month',
  YEAR: 'year'
}
// Adds leading zero to month/day if necessary, returns 'YYYY' if type = 'year',
// 'YYYY-MM' if 'month' and 'YYYY-MM-DD' if 'date'
export function sanitizeDateString(dateString/*: string*/, type/*: SANITY_TYPE*/)/*: string*/ {
  const [year, month = 1, date = 1] = dateString.split('-')
  return `${year}-${pad(month)}-${pad(date)}`.substr(0, { date: 10, month: 7, year: 4 }[type])
}
