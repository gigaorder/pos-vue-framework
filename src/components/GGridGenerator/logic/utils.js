import _ from 'lodash'
// References: https://css-tricks.com/snippets/css/complete-guide-grid/
export const _gridItemOptions = ['', 'start', 'end', 'center', 'stretch']
export const _gridContentOptions = ['', 'start', 'end', 'center', 'stretch', 'space-around', 'space-between', 'space-evenly']


/**
 * Join ref string
 * @param refArray
 * @param separator
 * @returns {string}
 */
export function joinRefArrayValue(refArray, separator = ' ') {
  let output = ''
  refArray.forEach(refVar => output += separator + refVar.value)
  return output
}

/**
 * An area created by 2 point. The position of these point is undetermined (top-left, top-right, bottom-left, bottom-right)
 * This method will convert an undetermined area to determined area which contain information of  top-left and bottom-right points
 * @param area
 * @returns {{columnEnd: number, columnStart: number, rowStart: number, rowEnd: number}}
 */
export function normalizeArea(area) {
  const rowStart = Math.min(area.rowStart, area.rowEnd) + 1
  const rowEnd = Math.max(area.rowStart, area.rowEnd) + 2
  const columnStart = Math.min(area.columnStart, area.columnEnd) + 1
  const columnEnd = Math.max(area.columnStart, area.columnEnd) + 2
  return {
    rowStart, rowEnd, columnStart, columnEnd
  }
}

export function getCssArea({rowStart, rowEnd, columnStart, columnEnd}) {
  return `${rowStart} / ${columnStart} / ${rowEnd} / ${columnEnd}`
}

