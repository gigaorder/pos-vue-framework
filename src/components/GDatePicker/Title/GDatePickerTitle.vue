<template>
  <div :class="titleClasses">
    <div :class="titleYearClasses" @click.stop="onYearButtonClicked">
      {{ year }}
    </div>
    <div :class="titleDateClasses">
      <transition :name="transitionName">
        <div :key="date" v-html="date"></div>
      </transition>
    </div>
  </div>
</template>

<script>
  import { computed } from '@vue/composition-api'
  import { TRANSITION_NAMES } from '../utils';
  import { getYearButtonClickHandler } from './GDatePickerTitleUtil'

  // class
  const PICKER_TITLE = 'g-date-picker-title'
  const PICKER_TITLE_DISABLED = 'g-date-picker-title--disabled'
  const PICKER_TITLE_BTN = 'g-picker__title__btn'
  const PICKER_TITLE_BTN_READONLY = 'g-picker__title__btn--readonly'
  const PICKER_TITLE_BTN_ACTIVE = 'g-picker__title__btn--active'
  const PICKER_TITLE_YEAR = 'g-date-picker-title__year'
  const PICKER_TITLE_DATE = 'g-date-picker-title__date'

  export default {
    name: 'GDatePickerTitle',
    props: {
      // 1st line -> year
      year: {
        type: [Number, String],
        default: ''
      },
      // 2nd line, maybe contain html code -> render as raw html
      date: {
        type: String,
        default: ''
      },
      // Boolean value indicate that this component is disabled or not
      disabled: Boolean,
      // Boolean value indicate that this component is readonly or not
      readonly: Boolean,
      // Boolean value indicate that whether Years is showing or not
      selectingYear: Boolean,
    },
    setup(props, context) {
      const titleClasses = computed((() => {
        return {
          [PICKER_TITLE]: true,
          [PICKER_TITLE_DISABLED]: props.disabled
        }
      }))

      // Year info
      const titleYearClasses = computed((() => {
        return {
          [PICKER_TITLE_BTN]: true,
          [PICKER_TITLE_BTN_READONLY]: props.readonly,
          [PICKER_TITLE_BTN_ACTIVE]: true,
          [PICKER_TITLE_YEAR]: true
        }
      }))

      const onYearButtonClicked = getYearButtonClickHandler(props, context)

      // Date/selected items state
      let titleDateClasses = computed(() => {
        return {
          [PICKER_TITLE_BTN]: true,
          [PICKER_TITLE_BTN_READONLY]: props.readonly,
          [PICKER_TITLE_DATE]: true
        }
      })

      return {
        titleClasses: titleClasses,
        //
        titleYearClasses: titleYearClasses,
        onYearButtonClicked,
        //
        titleDateClasses: titleDateClasses,
        //
        transitionName: TRANSITION_NAMES.PICKER
      }
    }
  }
</script>
