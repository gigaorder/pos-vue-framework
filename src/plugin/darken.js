import {getHslColor} from "../utils/colors";

;(function (window) {
  'use strict';

  const Darken = window.Darken || {};

  const Effect = {
    //click
    elements: [],
    // Effect delay
    duration: 1500,

    show: function (e, element) {

      // Disable right click
      if (e.button === 2) {
        return false;
      }

      const el = element || this;
      Effect.elements.push(el)

      const background = el.style.backgroundColor,
          border = el.style.borderColor,
          shadow = el.style.boxShadow

      let {h, s, l, a} = getHslColor(background)
      if (h === 0 && s === 0) {
        let hsla
        if (border)
          hsla = getHslColor(border)
        else
          hsla = {h: 0, s: 0, l: 0, a: 0}
        h = hsla.h
        s = hsla.s
        l = hsla.l
        a = hsla.a
      }
      el.style.background = `hsla(${h}, ${s - 5}%, ${l - 15}%, ${a})`
      el.style.borderColor = `hsla(${h + 2}, ${s - 10}%, ${l - 15}%, ${a})`
      el.style.boxShadow = `1px 0px 5px hsla(${h + 2}, ${s - 10}%, ${l - 15}%, ${a})`;

      setTimeout(() => {
        el.style.backgroundColor = background
        el.style.borderColor = border
        el.style.boxShadow = shadow
        const index = Effect.elements.findIndex(element => element === el)
        Effect.elements.splice(index, 1)
      }, Effect.duration)
    },

  };

  const TouchHandler = {
    touches: 0,
    allowEvent: function (e) {
      let allow = true;

      if (e.type === 'touchstart') {
        TouchHandler.touches += 1;
      } else if (e.type === 'touchend' || e.type === 'touchcancel') {
        setTimeout(function () {
          if (TouchHandler.touches > 0) {
            TouchHandler.touches -= 1;
          }
        }, Effect.duration);
      } else if (e.type === 'mousedown' && TouchHandler.touches > 0) {
        allow = false;
      }

      return allow;
    },
    touchup: function (e) {
      TouchHandler.allowEvent(e);
    }
  };

  function getWavesEffectElement(e) {
    if (TouchHandler.allowEvent(e) === false) {
      return null;
    }

    let element = null;
    let target = e.target || e.srcElement;

    while (target.parentNode !== null) {
      if (!(target instanceof SVGElement) && target.className.indexOf('darken-effect') !== -1) {
        element = target;
        break;
      }
      target = target.parentNode;
    }
    return element;
  }

  function showEffect(e) {
    const element = getWavesEffectElement(e);

    if (element !== null && Effect.elements.findIndex(el => el === element) === -1) {
      Effect.show(e, element);
    }
  }

  Darken.displayEffect = function (options) {
    options = options || {};

    if ('duration' in options) {
      Effect.duration = options.duration;
    }


    if ('ontouchstart' in window) {
      document.body.addEventListener('touchstart', showEffect, false);
    }

    document.body.addEventListener('mousedown', showEffect, false);
  };

  Darken.attach = function (element) {
    if ('ontouchstart' in window) {
      element.addEventListener('touchstart', showEffect, false);
    }

    element.addEventListener('mousedown', showEffect, false);
  };

  window.Darken = Darken;

  document.addEventListener('DOMContentLoaded', function () {
    Darken.displayEffect();
  }, false);

})(window);
