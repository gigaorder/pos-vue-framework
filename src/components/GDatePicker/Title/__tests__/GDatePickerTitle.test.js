import { createLocalVue, mount } from '@vue/test-utils'
import plugin from '@vue/composition-api'

import GDatePickerTitle from '../GDatePickerTitle'
import {EVENT_NAMES} from '../GDatePickerTitleUtil'
import {TRANSITION_NAMES} from '../../utils'

describe('GDatePickerTitle.js', () => {
  let mountFunction

  const localVue = createLocalVue()
  localVue.use(plugin)

  beforeEach(() => {
    mountFunction = (options) => {
      return mount(GDatePickerTitle, {
        localVue,
        ...options
      })
    }
  })

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        year: '1234',
        date: '2005-11-01',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render disabled component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        year: '1234',
        date: '2005-11-01',
        disabled: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render readonly component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        year: '1234',
        date: '2005-11-01',
        readonly: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component when selecting year and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        year: '1234',
        date: '2005-11-01',
        selectingYear: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit input event on year click', (done) => {
    const wrapper = mountFunction({
      propsData: {
        year: '1234',
        date: '2005-11-01',
      },
    })

    const input = jest.fn(value => wrapper.setProps({ selectingYear: value }))
    wrapper.vm.$on(EVENT_NAMES.UPDATE_SELECTING_YEAR, input)
    wrapper.findAll('.g-date-picker-title__year').at(0).trigger('click')
    expect(input).toHaveBeenCalledWith(true)
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.selectingYear).toBe(true)
      done()
    })
  })

  it('should have the correct transition', () => {
    const wrapper = mountFunction({
      propsData: {
        year: '2018',
        date: 'Tue, Mar 3',
        value: '2018-03-03',
      },
    })

    expect(wrapper.vm.transitionName).toEqual(TRANSITION_NAMES.PICKER)
  })
})
