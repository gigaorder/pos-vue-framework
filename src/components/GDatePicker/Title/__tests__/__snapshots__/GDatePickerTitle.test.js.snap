// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GDatePicker/Title/GDatePickerTitle.js should render component and match snapshot 1`] = `
<div class="g-date-picker-title">
  <div class="g-picker__title__btn g-picker__title__btn--active g-date-picker-title__year">
    1234
  </div>
  <div class="g-picker__title__btn g-date-picker-title__date">
    <div name="picker-transition">
      2005-11-01
    </div>
  </div>
</div>
`;

exports[`GDatePicker/Title/GDatePickerTitle.js should render component when selecting year and match snapshot 1`] = `
<div class="g-date-picker-title">
  <div class="g-picker__title__btn g-picker__title__btn--active g-date-picker-title__year">
    1234
  </div>
  <div class="g-picker__title__btn g-date-picker-title__date">
    <div name="picker-transition">
      2005-11-01
    </div>
  </div>
</div>
`;

exports[`GDatePicker/Title/GDatePickerTitle.js should render disabled component and match snapshot 1`] = `
<div class="g-date-picker-title g-date-picker-title--disabled">
  <div class="g-picker__title__btn g-picker__title__btn--active g-date-picker-title__year">
    1234
  </div>
  <div class="g-picker__title__btn g-date-picker-title__date">
    <div name="picker-transition">
      2005-11-01
    </div>
  </div>
</div>
`;

exports[`GDatePicker/Title/GDatePickerTitle.js should render readonly component and match snapshot 1`] = `
<div class="g-date-picker-title">
  <div class="g-picker__title__btn g-picker__title__btn--readonly g-picker__title__btn--active g-date-picker-title__year">
    1234
  </div>
  <div class="g-picker__title__btn g-picker__title__btn--readonly g-date-picker-title__date">
    <div name="picker-transition">
      2005-11-01
    </div>
  </div>
</div>
`;
